simplified/HordeAd/Core/Ast.hs

```diff
@@ -20,30 +23,174 @@ module HordeAd.Core.Ast
 import Prelude
 
 import           Control.Exception.Assert.Sugar
+import           Data.Array.Internal (valueOf)
 import qualified Data.Array.RankedS as OR
 import           Data.IORef.Unboxed (Counter, atomicAddCounter_, newCounter)
 import           Data.Kind (Type)
+import           Data.Monoid (All (..))
 import           Data.MonoTraversable (Element, MonoFunctor (omap))
 import qualified Data.Strict.Vector as Data.Vector
 import qualified Data.Vector.Generic as V
 import qualified Data.Vector.Storable as VS
 import           GHC.TypeLits (KnownNat, Nat, type (+))
-import           Numeric.LinearAlgebra (Numeric)
+import           Numeric.LinearAlgebra (Numeric, Vector)
 import           System.IO.Unsafe (unsafePerformIO)
 import           Text.Show.Functions ()
+import           Unsafe.Coerce (unsafeCoerce)
 
 import HordeAd.Internal.OrthotopeOrphanInstances ()
 
--- * Definitions
+-- * GHC.Nat-indexed lists, originally by Tom Smeding
+
+-- This needs to be moved to a better place, probably its own module,
+-- once its clear which other modules depend on it.
+
+-- | An index in an n-dimensional array. The fastest-moving index is at the
+-- last position; thus the index 'Z :. i :. j' represents 'a[i][j]' in
+-- traditional C notation.
+data Index (n :: Nat) i where
+  Z :: Index 0 i
+  S :: i -> Index n i -> Index (n + 1) i
+
+deriving stock instance Functor (Index n)
+
+instance Show i => Show (Index n i) where
+  showsPrec _ Z = showString "Z"
+  showsPrec d (idx :. i) = showParen (d > 3) $
+    showsPrec 3 idx . showString " :. " . showsPrec 4 i
+
+-- I'm afraid, I can't do the unsafeCoerces below with this order
+-- of argument in :., can I? So I need this instead:
+pattern (:.) is i = S i is
```

Comment:
@tomsmeding, at your leisure, if you could have  alook at 2 problems I've encountered so far and my partial solutions. This is the first.

Reply:
I get what you are doing with the unsafeCoerces, and that's horrible. :D In particular I'd ask you to add a strongly worded comment above the definition of `Index` imploring the reader to not change the representation of the data type; for example, adding an `n >= 0` constraint to the `(:.)` case is already wrong if I'm not mistaken, because that becomes an additional field in the data type, breaking representational equality with lists. Or so I've been told — happy to be corrected here.

And no, if you want representational equality with lists, then you cannot change the, well, representation, so no way around using a pattern synonym to fix the argument order. :)

Reply:
Agreed. I will try to add a strongly worded comment that is still family-friendly.

Reply:
I wonder if you really need the unsafeCoerces, though. `take` on a list js not going to be faster than reimplementing the same directly on `Index`; indeed, as we've determined, the two have the same representation so hopefully we'll even get the exact same code out in the end. Same for `tail` and `drop`.

For permutations, `VS.fromList` has to walk the list anyway, so reimplementing that for `Index` should again result in the same code. Then, `VS.toList` could be reimplemented by just indexing into the output vector — that should be the same as what fromList is doing now.

And here we have eliminated all unsafeCoerce, I believe, meaning that you can change the argument order of the second constructor of `Index` back to what it should be! :)

Reply:
Haha, you've elided the tiny detail of actually writing `take` without `unsafeCoerce` using non-unary Nat. And the same for any future list function we realise we need. I did try to write `take` in a honest way first...

Reply:
Do you happen to know why [this thing](https://hackage.haskell.org/package/base-4.16.0.0/docs/src/GHC.TypeNats.html#withSNat) is not exported? It would be nice here.

Also did I mention that I hate GHC Nats?

But you're right of course. Still I wonder whether the following would incur any runtime penalty:
```haskell
takeIndex :: forall len n i. KnownNat len
          => Index (len + n) i -> Index n i
takeIndex idx =
  let len = natVal (Proxy @len)
  in if len < 0 then error "wat" else go len idx
  where
    go :: Integer -> Index m i -> Index m' i
    go 0 idx' = unsafeCoerce idx'
    go len Z = error "wat"
    go len (S _ idx') = go (len - 1) idx'
```
If not, that would allow at least avoiding the `(:.)` pattern synonym (because we can fix the argument order in the data type).

Reply:
> Do you happen to know why [this thing](https://hackage.haskell.org/package/base-4.16.0.0/docs/src/GHC.TypeNats.html#withSNat) is not exported? It would be nice here.

No, I don't know. Probably authors thought this bit is less stable than the rest. But if so, it should be in an .EvenMore.Internal module. Perhaps they can be persuaded `withSNat` itself seem pretty solid and essential, unlike the hacks used to implement it. Analogous things are certainly provided and exposed for other Nat implementations.

Re `nonsense`, where would the runtime penalty come from? I don't see anything suspect.

Re `(:.)` pattern synonym, perhaps let's keep it? I have a dream that one day we'd have such a sized list in base and then I'd like to just effortlessly replace our implementation with it, keeping the synonym and a few specialized functions (and `Shape`).

Reply:
Hm, fair point, perhaps having the pattern synonym doesn't really bring any downsides. (I don't have particularly high hopes for a suitable sized list type in base, but who knows.)

I also don't think `nonsense`-style functions would bring a runtime penalty, but you have more knowledge of such matters than I do, hence I asked :)
The advantage of this style is being independent of the fragile equality of representation between `Index n` and `[]`; it doesn't save any `unsafeCoerce`. The disadvantage is clearly verbosity.

Reply:
(My knowledge here comes from writing an early incorrect version of, and later understanding the actually correct version of, [this code here](https://github.com/AccelerateHS/accelerate/blob/5971c5d8e4dbba28d2017e7ce422cf46a20197cb/src/Data/Array/Accelerate/AST/Idx.hs#L65-L92))

Reply:
Turns out other people thought the same about `SNat` and friends, so [in GHC 9.6](https://github.com/haskell/core-libraries-committee/issues/85) this will change :)

Reply:
Oh, cool. Do I understand correctly that this would enable a purer implementation of some or our functions, but it does not break our dirty code, at least not yet in GHC 9.6?